design/baremetal-operator/bmh_non-bootable_iso.md

```diff
@@ -0,0 +1,124 @@
+has been Provisioned or ExternallyProvisioned.
+
+### Ironic Support
+
+The [proposal](https://bugs.launchpad.net/ironic/+bug/2033288) for exposing this functionaility in Ironic has been
+accepted by the upstream Ironic community.
+
+The [Ironic implementation](https://review.opendev.org/c/openstack/ironic/+/894918) based on the above proposalon has been completed.
+
+Other than that, we just need the implementation for driver specific support,
+which in this proposal is Redfish.
+
+## Design Details
+
+Add an optional section `DataImage` to `BareMetalHostSpec`, with sub-field
+`Url`.
+
+The ISO will be attached when the `BareMetalHost` object is edited to add
+the `DataImage` spec triggering a reconciliation. In case a reboot is requested
+at the same time using `RebootAnnotation`, the ISO will be attached first.
+
+The `BareMetalHostStatus` will cache the image details, which will inform
+us if the attachment succeeded and will also be used for detachment, either
+when the attachment fails and we retry or when the user explicitly requests
+detachment by removing the `DataImage` spec from `BareMetalHost` object.
+
+A flag will be used at the webhook level to validate if a given driver supports the
+non-bootable ISO feature. Since the support for attaching an ISO as virtual media is
+equivalent for both `PreprovisioningISOImage` and `NonBootableISOImage`, we plan to
+rename the existing flag for `PreprovisioningISOImage` i.e. `SupportsISOPreprovisioningImage`
+to `SupportVirtualMediaISOImage` and use it for both the cases.
+
+### Implementation Details
+
+As part of the design we are assuming that the user will edit the BareMetalHost
+to add the dataImage spec, and might also request reboot via rebootAnnotation
+or power state change via `online` field.
+
+We want to attach the dataImage first, so handling the attachment/detachment of dataImage
+inside the Reconcile function, after reconciling host data and updating hardware details ([code reference](https://github.com/metal3-io/baremetal-operator/blob/1bb45eef449c942711b1c0937ecff2b10a326eb3/controllers/metal3.io/baremetalhost_controller.go#L152) ), makes sense since
+any power state change (including reboot via rebootAnnotation) will be handled later on via the stateMachine ( [code reference](https://github.com/metal3-io/baremetal-operator/blob/1bb45eef449c942711b1c0937ecff2b10a326eb3/controllers/metal3.io/baremetalhost_controller.go#L222) ).
+
+We need to cache the attached image in `Status` of the BareMetalHost to know if the
+image has been attached. This will also help us determine if a different image is
+attached previously which needs to be detached first before attaching the new
+dataImage. Or if the user delets the dataImage spec and the `Status` contains a
+cached entry, which will again trigger detachment of that image.
+
+In case the image fails to attach, we will retry until either the attachment succeeds
```

Comment:
Will this cause the state of the BMH to go to "provisioning error" or would this be a different error state? or would be in general considered a blocking error?

Reply:
The attachment of detachment of DataImage will happen during reconciliation of the BMH, so in case of a failure it will raise an error in the logs and requeue the request similar to what we do when [fetching the host here](https://github.com/metal3-io/baremetal-operator/blob/63abe2db90306baa805680ec391b055d9fe5be31/controllers/metal3.io/baremetalhost_controller.go#L115).